tools/walletextension/ratelimiter/rate_limiter.go

```diff
@@ -0,0 +1,123 @@
+package ratelimiter
+
+import (
+	"log"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Package ratelimiter implements a simple rate limiting mechanism
+// using a score-based approach. Each user has a score that decays
+// over time. Requests are allowed if the user's score is below a
+// specified threshold. The score increases with each request and
+// decays based on the time since the last request.
+
+type Score struct {
+	lastRequest time.Time
```

Comment:
let's add a "concurrentRequests"  value here as well, which we increment during `Allow` and decrement during "Update".

In "Allow" we should also check that the concurrentRequests needs to be below a configured threshold (A small value like 3).

With this mechanism we protect someone from Dos-ing the gateway by triggering a very large number of expensive requests

Reply:
added